### PR TITLE
Cleaned up expressions

### PR DESCRIPTION
**Pull request by jarshwah**
Fixes some cruft that was accidentally left in. Adds repr methods to the majority of expressions. Renames CombinableMixin to Combinable.

I was also going to add a commit to introduce the concept of Resolvable, but it was unclear which methods should be moved to that type, and it unnecessarily complicated the implementation so I ditched the idea.

I can split the changes here into multiple commits if need be, and add tickets for each one also.

### CODE REVIEW

File: django/db/models/aggregates.py [hunk 1]
```diff
@@ -105,6 +105,12 @@ def __init__(self, expression, distinct=False, **extra):
         super(Count, self).__init__(
             expression, distinct='DISTINCT ' if distinct else '', output_field=IntegerField(), **extra)
 
+    def __repr__(self):
+        args = self.arg_joiner.join(str(arg) for arg in self.source_expressions)
+        if self.extra['distinct'] == '':
+            return "{}({}, distinct=False)".format(self.__class__.__name__, args)
+        return "{}({}, distinct=True)".format(self.__class__.__name__, args)
```
Comment (timgraham) [inline django/db/models/aggregates.py]: Is there a reason you didn't opt for something like this:
```
return "{}({}, distinct={})".format(
    self.__class__.__name__,
    self.arg_joiner.join(str(arg) for arg in self.source_expressions),
    'False' if self.extra['distinct'] == '' else 'True',
)
```
(to repeat less code)

File: django/db/models/expressions.py [hunk 2]
```diff
@@ -184,6 +184,7 @@ def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize
            in this query
          * reuse: a set of reusable joins for multijoins
          * summarize: a terminal aggregate clause
+         * for_save: is this expression about to be used in a save or update
```
Comment (timgraham) [inline django/db/models/expressions.py]: "whether this expression is about to be used..."
(or add a question mark at the end if you prefer the "is" version)

File: django/db/models/expressions.py [hunk 3]
```diff
@@ -571,6 +594,10 @@ def __init__(self, alias, target, source=None):
         super(Col, self).__init__(output_field=source)
         self.alias, self.target = alias, target
 
+    def __repr__(self):
+        return "{}({}, {})".format(
```
Comment (timgraham) [inline django/db/models/expressions.py]: one line should be okay here

File: tests/expressions/tests.py [hunk 4]
```diff
@@ -812,3 +819,41 @@ def test_update_UUIDField_using_Value(self):
...
+    def test_functions(self):
+        self.assertEqual(repr(Coalesce('a', 'b')), "Coalesce(F(a), F(b))")
+        self.assertEqual(repr(Concat('a', 'b')), "Concat(ConcatPair(F(a), F(b)))")
+        self.assertEqual(repr(Length('a')), "Length(F(a))")
+        self.assertEqual(repr(Lower('a')), "Lower(F(a))")
+        self.assertEqual(repr(Substr('a', 1, 3)), "Substr(F(a), Value(1), Value(3))")
+        self.assertEqual(repr(Upper('a')), "Upper(F(a))")
+
+    def test_aggregates(self):
+        Avg, Count, Max, Min, StdDev, Sum, Variance
```
Comment (timgraham) [inline tests/expressions/tests.py]: remove?

**Comment by timgraham** [comment]
Separate commits would be helpful:
1. the bug fix and the `test_reverse_ordering_pure` tests
2. repr additions
3. anything else?
Referencing the original ticket should probably be fine.

**Comment by jarshwah** [comment]
I've separated the commits into logical chunks. I feel each is good to go. Any comments before I start merging?

**Comment by timgraham** [comment]
Just double checking the test is supposed to be alone. I thought maybe the descending_template/ ascending_template removals were related. A slightly clearer commit message might be something like "Added a test to demonstrate that reverse order only affects the current queryset", otherwise LGTM.

**Comment by jarshwah** [comment]
I thought there was going to be a bug, so I wrote a test to check. No bug though. I'll update the commit message in the morning and push it through.

The asc/desc_template was cruft I was going to use but decided against.

**Pull request merged**